src/main/java/io/cryostat/discovery/DiscoveryStorage.java

```diff
@@ -65,18 +75,100 @@ public class DiscoveryStorage extends AbstractPlatformClientVerticle {
+            Path persistencePath,
+            Gson gson,
+            WebClient http,
+            Logger logger) {
         this.uuid = uuid;
+        this.fs = fs;
+        this.persistencePath = persistencePath;
+        this.gson = gson;
+        this.http = http;
         this.logger = logger;
     }
 
     @Override
-    public void start() throws Exception {
-        // TODO persist plugin infos (with empty subtrees) to disk on shutdown, and reinitialize map
-        // here. Then, perform POST on each callback URI to check it's still there and prompt it to
-        // update us with its subtree
+    public void start(Promise<Void> future) throws Exception {
+        for (String s : fs.listDirectoryChildren(persistencePath)) {
+            Path p = persistencePath.resolve(s);
+            try (BufferedReader br = fs.readFile(p)) {
+                PluginInfo pluginInfo = gson.fromJson(br, PluginInfo.class);
+                if (Objects.equals(pluginInfo.getCallback(), NO_CALLBACK)) {
+                    continue;
+                }
+                map.put(UUID.fromString(s), pluginInfo);
+            } catch (IOException ioe) {
+                logger.error(ioe);
+            }
+        }
+
+        map.entrySet()
+                .forEach(
+                        plugin -> {
+                            UUID key = plugin.getKey();
+                            URI uri = plugin.getValue().getCallback();
+                            http.postAbs(uri.toString())
+                                    .timeout(1_000)
+                                    .followRedirects(true)
+                                    .send()
+                                    .onSuccess(
+                                            res -> {
+                                                if (!HttpStatusCodeIdentifier.isSuccessCode(
+                                                        res.statusCode())) {
+                                                    removePlugin(key, uri);
+                                                    return;
+                                                }
+                                                logger.info("Found discovery service {}", uri);
+                                            })
+                                    .onFailure(
+                                            t -> {
+                                                t.printStackTrace();
+                                                removePlugin(key, uri);
+                                            });
+                        });
+        future.complete();
+    }
+
+    private void removePlugin(UUID uuid, URI uri) {
+        map.remove(uuid);
+        try {
+            fs.deleteIfExists(persistencePath.resolve(uuid.toString()));
+        } catch (IOException ioe) {
+            logger.error(ioe);
+        }
+        logger.info("Stale discovery service {} removed", uri);
+    }
+
+    @Override
+    public void stop() {
```

Comment:
I wonder in the case of immediate kill like `SIGKILL`, does this method also get called to clean up or will we lose all discovery tree? Not sure I word that right :D

Reply:
Nope, if the OS `SIGKILL`s you then by definition you don't get to do any clean up, and anything that was only in-memory is lost.

Reply:
Are we planning to do a backup by some intervals? Or this loss is okay?

Reply:
In #1031 with the real database, every update sent to us is handled by a database transaction that persists changes as they happen. No intervals needed, and losses are about as minimized as they can be.

Reply:
Oh okayy understood! I guess I will check out PR with the real database.

Reply:
This file-based persistence PR is really just an implementation stub stepping toward #1031 anyway - the file backing is there just so I could implement and test the storage startup behaviour where it checks for already-known callback URLs and tests them.

We have other places that also use file-based storage and update that per-request too, like the JMX Credentials storage, so no interval-based strategy is needed either.

Reply:
Well, don't get too far ahead on #1031. That one is still in progress. This one is meant to be fleshing out the client-facing API and internal implementation for the behaviours, but the file-based storage is a temporary measure as a stand-in for a database.